### ISSUE TITLE
Error handling and errorCallback

### ISSUE THREAD

**Issue by garyscavone** [thestk/rtaudio#195]
In the course of trying to address issue #194, I ran into error handling problems. For that issue, the goal was to automatically close an open stream if the device is removed and report an error to the user. The original error functionality in RtAudio relied on C++ exceptions and the code was written with the expectation that if an error occurred, an exception would be thrown (perhaps by the verifyStream() function) and processing would stop. For example, if you call stopStream() when the stream is closed, an exception should be thrown right away, otherwise segmentation faults will occur during attempts to access non-existent memory and data structures. The rationale for the use of the exceptions was that any subsequent actions after the corresponding errors c/would result in undefined audio system behaviour.

One problem can occur when using an errorCallback function. In this case, it is very possible that the verifyStream() function returns, even after an error, and the rest of the function is processed (using the example above, the code attempts to stop a stream that is already closed, which fails). I like the option of an errorCallback but it makes it harder to block execution of subsequent code.

A see a few possible fixes:
     1. Add an extra check after the verifyStream() calls to make sure the stream is open and simply return if not. That would require quite a lot of updates to the code but it should be relatively simple (cut and paste). That said, the user could end up with a system that doesn't abort but doesn't work either.
     2. Only invoke the errorCallback for warnings (and throw exceptions for critical errors).
     3. Reduce the dependence on exceptions and go with the more traditional approach of using a return value to signal an error condition (at least for functions like openStream(), startStream(), stopStream(), and abortStream() that don't currently return a value). This would entail a significant code rewrite, including for the wrappers. Warning or error messages would still be generated but the user would have to verify function success at each step.

Other ideas or preferences?

**Comment by ad8e** [comment]
When I tried RTAudio, PortAudio, and libsoundio, I found that RTAudio used exceptions and iostreams. I had avoided these two features in my own C++ code, and I didn't want to bring them in, so I decided on PortAudio. However, exceptions are sometimes preferred by other C++ programmers, especially academic programmers. Since RTAudio seems to be more academically oriented (?), and doesn't have large resources to throw at maintenance (like the WASAPI backend), it may not be worth removing exceptions.

In short: 

1. I don't like exceptions in libraries, but some people might.
2. The effort maybe isn't worth it for you, judging by the time you have available to work on this.
3. In my codebase, which supports both PortAudio and RTAudio, handling error conditions was a simple matter of wrapping all library calls: pa_run(PA_Function()).

**Comment by renaud-sc** [comment]
I would not call myself a C++ guru but to me exceptions doesn't fit very well in a multithread/async environment. So I would choose option 3. But I understand it is a major work an more likely an API breaker. So option 3 is probably a long term goal with major version bump while option 1 is probably the best option on a short term. 
Option 2 is really not an option ;-) a correct error handling which would at least stop the stream and offer callback with a simple error message (to be used for GUI or logging purpose) is really something I am waiting for :-)

**Comment by garyscavone** [comment]
I'm closing this, as we have now removed exceptions.
